File: mRemoteV1/Controls/TextBox.vb
Imports System.ComponentModel

' Adapted from http://stackoverflow.com/a/3678888/2101395

Namespace Controls
    Public Class TextBox
        Inherits System.Windows.Forms.TextBox

#Region "Public Properties"

        <Category("Behavior"),
            DefaultValue(False)>
        Public Property SelectAllOnFocus As Boolean = False

#End Region

#Region "Protected Methods"

        Protected Overrides Sub OnEnter(e As EventArgs)
            MyBase.OnEnter(e)

            If MouseButtons = MouseButtons.None Then
                SelectAll()
                _focusHandled = True
            End If
        End Sub

        Protected Overrides Sub OnLeave(e As EventArgs)
            MyBase.OnLeave(e)

            _focusHandled = False
        End Sub

        Protected Overrides Sub OnMouseUp(e As MouseEventArgs)
            MyBase.OnMouseUp(e)

            If Not _focusHandled Then
                If SelectionLength = 0 Then SelectAll()
                _focusHandled = True
            End If
        End Sub

#End Region

#Region "Private Fields"

        Private _focusHandled As Boolean = False

#End Region
    End Class
End Namespace

File: mRemoteV1/UI/Forms/TextBox.cs
using System;
using System.ComponentModel;
using System.Windows.Forms;

// Adapted from http://stackoverflow.com/a/3678888/2101395

namespace mRemoteNG.UI.Forms
{
	public class TextBox : Controls.Base.NGTextBox
	{
        #region Public Properties
		[Category("Behavior"),
			DefaultValue(false)]private bool _SelectAllOnFocus;
        public bool SelectAllOnFocus
		{
			get
			{
				return _SelectAllOnFocus;
			}
			set
			{
				_SelectAllOnFocus = value;
			}
		}
        #endregion
			
        #region Protected Methods
		protected override void OnEnter(EventArgs e)
		{
			base.OnEnter(e);
				
			if (MouseButtons == MouseButtons.None)
			{
				SelectAll();
				_focusHandled = true;
			}
		}
			
		protected override void OnLeave(EventArgs e)
		{
			base.OnLeave(e);
				
			_focusHandled = false;
		}
			
		protected override void OnMouseUp(MouseEventArgs e)
		{
			base.OnMouseUp(e);
				
			if (!_focusHandled)
			{
				if (SelectionLength == 0)
				{
					SelectAll();
				}
				_focusHandled = true;
			}
		}
        #endregion
			
        #region Private Fields
		private bool _focusHandled;
        #endregion
	}
}